Lesion segmentation protocol (#9)
ivadomed/ms-lesion-agnostic: lesion_segmentation_protocol.md

# Lesion segmentation protocol:

The following details the protocol for Multiple Sclerosis (MS) lesion segmentation in the spinal cord.
Imaging the spinal cord is often essential to confirm the diagnosis of MS. That is because the lesions of the spinal cord are included in the McDonald diagnosis criteria, which studies dissemination in space and in time [(Thompson et al. 2018)](https://pubmed.ncbi.nlm.nih.gov/29275977/). While the MAGNIMS-CMSC-NAIMS working group recommends to use at least two sagittal images for MS diagnosis, still, axial imaging is mentioned as optional in international imaging guidelines [(Wattjes et al. 2021)](https://pubmed.ncbi.nlm.nih.gov/34139157/).

> For detecting MS lesions in the spinal cord, two main contrasts emerge: PSIR and STIR contrasts. New studies [(Peters et al. 2024)](https://pubmed.ncbi.nlm.nih.gov/38289376/)[(Fechner et al. 2019)](https://pubmed.ncbi.nlm.nih.gov/30679225/) showed that using PSIR contrasts improved MS lesion detection in the spinal cord. [(Fechner et al. 2019)](https://pubmed.ncbi.nlm.nih.gov/30679225/) showed that the PSIR contrast showed a higher signal-to-noise (SNR) ratio compared to the STIR contrast. 

Reviewer: Maybe my suggestion is not up-to-date, but I have doubts/questions abouts STIR sequences being frequently used for assessing MS lesions, at least in standard/basics protocols (even if mentioned in the picture showing example of of clinically used sequences for MS). From my understanding, the most useful sequence for detecting MS lesions in clinical practice is proton density (PD), but probably that changes from sites/radiologists perspective.
<img width="339" alt="basic_sequences" src="https://github.com/ivadomed/ms-lesion-agnostic/assets/142258732/afe5cd16-69bd-43e1-b4a2-a3e9b7c4a3ff">
According to the articles, PSIR sequences seem great! However, here are some potential precisions I would address :
1-Peters and al : PSIR compared to STIR (not the best...) and T2;
2-Fechner : PSIR compared to T2 and T1C+ sequences (not compared to PD sequences...)
```suggestion
For detecting MS lesions in the spinal cord, two main contrasts emerge: PSIR and STIR contrasts. New studies [(Peters et al. 2024)](https://pubmed.ncbi.nlm.nih.gov/38289376/)[(Fechner et al. 2019)](https://pubmed.ncbi.nlm.nih.gov/30679225/) showed that using PSIR contrasts improved MS lesion detection in the spinal cord. [(Fechner et al. 2019)](https://pubmed.ncbi.nlm.nih.gov/30679225/) showed that the PSIR contrast showed a higher signal-to-noise (SNR) ratio compared to the STIR contrast. 
```

> ## Criteria to segment MS lesions in the spinal cord:

Reviewer: ```suggestion
## Criteria to segment MS lesions in the spinal cord:
```
I would add a clear definition of what consists a lesion according to the sequences you are using (in effort for standardization)

Author: This is a great idea ! Could you propose a definition ?

Reviewer: ```suggestion
## Criteria to segment MS lesions in the spinal cord:
Based on the definition proposed for MS lesions in the spinal cord in the 2017 revisions of the
McDonald criteria:
"A hyperintense lesion in the cervical, thoracic, or lumbar spinal cord seen on T2 plus short tau inversion recovery, proton-density images, or other appropriate sequences, or in two planes on T2 images."  [(Thompson et al. 2018)](https://pubmed.ncbi.nlm.nih.gov/29275977/) [(Filippi et al. 2016)](https://pubmed.ncbi.nlm.nih.gov/26822746/) [(Brownlee et al. 2018)](https://pubmed.ncbi.nlm.nih.gov/27889190/) [(Rovira et al. 2016)](https://pubmed.ncbi.nlm.nih.gov/26149978/) 
```

Author: Definition added. !

> Do not segment lesions in images with too many artifacts (such as this one : https://github.com/ivadomed/canproco/issues/53#issue-1938136790). Preferably, add the image to the exclude file so that it isn’t used for model training…

Reviewer 2: add hyperlink to exclude file

Author: Done !

> Do not segment lesions in images with too many artifacts (such as this one : https://github.com/ivadomed/canproco/issues/53#issue-1938136790). Preferably, add the image to the exclude file so that it isn’t used for model training…

Reviewer: Question : in words, what would be an image considered being too artefacted? Any criteria?
```suggestion
Do not segment lesions in images with too many artifacts (such as this one : https://github.com/ivadomed/canproco/issues/53#issue-1938136790). Preferably, add the image to the exclude file so that it isn’t used for model training…
```

When segmenting lesions on thick slices, always look at the above/below slices to build the volume of the lesion (this can minimize partial volume effect).
Do not segment lesions above the first vertebrae (because here we focus only on MS lesions in the spinal cord). 

> For lesions segmentations which you are not 100% sure, flag the subject and report it for external validation of the segmentation.

Reviewer: ```suggestion
If you have any doubt and/or are not 100% confident about one(or more) lesion(s) segmentation(s), flag the subject and report it for external validation of the segmentation.
```

Author: Done !

## How to manually segment lesions:

> - MS spinal cord lesions can be manually corrected from the prediction of a segmentation model or manually segmented from scratch. In the first case, make sure to build the json file associated with the segmentation prediction such as detailed in this comment (https://github.com/ivadomed/canproco/issues/73#issuecomment-1912302223). 

Reviewer: ```suggestion
- MS spinal cord lesions can be manually corrected from the prediction of a pre-existing segmentation model or manually segmented from scratch. In the first case, make sure to build the json file associated with the segmentation prediction such as detailed in this comment (https://github.com/ivadomed/canproco/issues/73#issuecomment-1912302223). 
```

Reviewer 2: @maxradx please make sure to make the comment on the latest commit, otherwise it cannot be applied as it is outdated:
![image](https://github.com/ivadomed/ms-lesion-agnostic/assets/2482071/2726edc3-3ef9-414b-980e-77bfa21990c6)

@plbenveniste I suggest you go through all outdated comments, and apply them/respond to them, and mark them as resolve (and hide them) so they don't clutter our discussion, as I did here:

![image](https://github.com/ivadomed/ms-lesion-agnostic/assets/2482071/a1635848-b3c9-446d-a6f2-a261ef254ed0)

Reviewer: ok !

Author: On it !

Author: Done !

> - MS spinal cord lesions can be manually corrected from the prediction of a segmentation model or manually segmented from scratch. In the first case, make sure to build the json file associated with the segmentation prediction such as detailed in this comment (https://github.com/ivadomed/canproco/issues/73#issuecomment-1912302223). 

Author: Good point ! And it is supposed to evolve as soon as Canproco is reformated to our new bids convention ! This is waiting for the data to be updated with Erin's QCed data

- For manual correction of the segmentation file use the manual-correction (https://github.com/spinalcordtoolbox/manual-correction) repository. The command can be inspired from this: 

```console
python manual_correction.py -path-img ~/data/canproco -config ~/config_seg.yml  -path-label ~/data/canproco/derivatives/labels  -suffix-files-lesion _lesion-manual -fsleyes-dr="-40,70"
``` 

> - Then a QC should be produced (prefarably using SCT) and added to a Github issue for further validation by other investigators. 

Reviewer 2: ```suggestion
- A Quality Control (QC) report should be produced using SCT, and added to a Github issue for further validation by other investigators. 
```

Author: Should we add the future commands that will be used by SCT for QCing lesion segmentations ? 
Because the current method is quite tricky and not so cleean. It requires to first segment the spinal cord and use sct_deepseg_lesion on sagittal plane to QC lesion segmentations.

Author: Discussion with Julien : 
- add current QC report
- open issue to update the QC report and the json standard

Author: Issue opened #13

Author: sct_qc commands added

> - If you are not sure of a subject, it should be flagged on Github for a more open discussion. 

Reviewer 2: link to 1-2 examples

> # Lesion segmentation protocol:

Reviewer 2: ```suggestion
# Lesion segmentation protocol
```

Author: Done!

For detecting MS lesions in the spinal cord, two main contrasts emerge: PSIR and STIR contrasts. New studies [(Peters et al. 2024)](https://pubmed.ncbi.nlm.nih.gov/38289376/)[(Fechner et al. 2019)](https://pubmed.ncbi.nlm.nih.gov/30679225/) showed that using PSIR contrasts improved MS lesion detection in the spinal cord. [(Fechner et al. 2019)](https://pubmed.ncbi.nlm.nih.gov/30679225/) showed that the PSIR contrast showed a higher signal-to-noise (SNR) ratio compared to the STIR contrast. 

## Criteria to segment MS lesions in the spinal cord:

- Do not segment lesions in images with too many artifacts (such as this [example](https://github.com/ivadomed/canproco/issues/53#issue-1938136790)). Preferably, add the image to the exclude file so that it isn’t used for model training…

> - When segmenting lesions on thick slices, always look at the above/below slices to build the volume of the lesion (this can minimize partial volume effect).

Reviewer 2: ```suggestion
- When segmenting lesions on thick slices, always look at the adjacent slices, as partial volume effect can sometimes reduce the appearance of a lesion (close to noise level).
```

Author: Done !

> - Do not segment lesions above the first vertebrae (because here we focus only on MS lesions in the spinal cord). 

Reviewer 2: ```suggestion
- Unless otherwise stated, do not segment lesions above the first vertebrae (because here we focus only on MS lesions in the spinal cord). 
```

Author: Done !

- For lesions segmentations which you are not 100% sure, flag the subject and report it for external validation of the segmentation.

> ## How to manually segment lesions:

Reviewer 2: ```suggestion
## How to manually segment lesions
```

> - MS spinal cord lesions can be manually corrected from the prediction of a segmentation model or manually segmented from scratch. In the first case, make sure to build the json file associated with the segmentation prediction such as the following :

Reviewer 2: ```suggestion
- MS spinal cord lesions can either (i) be automatically segmented from an algorithm and then manually corrected, or (ii) manually segmented from scratch. In the former case, make sure to use the JSON file that was created by the automatic segmentation algorithm, in order to track provenance:
```

Author: Done !

```json
{
        "GeneratedBy": [
            {
                "Name": "2D nnUNet model model_ms_seg_sc-lesion_regionBased.zip",
                "Version": "https://github.com/ivadomed/canproco/releases/tag/r20240125",
                "Date": "2024-01-26"
            }
        ]
    }
```

> - For manual correction of the segmentation file use the manual-correction (https://github.com/spinalcordtoolbox/manual-correction) repository. The command can be inspired from this: 

Reviewer 2: ```suggestion
- To manually create/correct the segmentation, please use the manual-correction (https://github.com/spinalcordtoolbox/manual-correction) repository. The command can be inspired from this: 
```

Author: Done !

```console
python manual_correction.py -path-img ~/data/canproco -config ~/config_seg.yml  -path-label ~/data/canproco/derivatives/labels  -suffix-files-lesion _lesion-manual -fsleyes-dr="-40,70"
``` 

> - A Quality Control (QC) report should be produced using SCT, and added to a Github issue for further validation by other investigators. Using SCT, you can review lesion segmentation in the axial or sagittal plane :

Reviewer 2: ```suggestion
- A Quality Control (QC) report should be produced using SCT, and added to a Github issue for further validation by other investigators. Using SCT, you can review lesion segmentation in the axial or sagittal plane:
```

```console
sct_qc -i {image_file} -d {lesion_seg_file} -s {sc_seg_file} -p sct_deepseg_lesion -plane {sagittal, axial} -qc {canproco_qc_folder}
```  

> - If you are not sure of a subject, it should be flagged on Github for a more open discussion: here are some examples [(1)](https://github.com/ivadomed/ms-lesion-agnostic/issues/4#issuecomment-1947326493) and [(2)](https://github.com/ivadomed/ms-lesion-agnostic/issues/4#issuecomment-1947338624)

Reviewer 2: ```suggestion
- If you are not sure about the segmentation on a subject, it should be flagged on GitHub for a more open discussion: here are some examples [(1)](https://github.com/ivadomed/ms-lesion-agnostic/issues/4#issuecomment-1947326493) and [(2)](https://github.com/ivadomed/ms-lesion-agnostic/issues/4#issuecomment-1947338624)
```

Author: Done !

> ## Step 1: Get familiar with FSLeyes and SCT:

Reviewer 2: ```suggestion
## Software to create/correct lesions
```

Author: Title changed to `Step 1: Software to create/correct lesions`

It is common practice to use FSLeyes at NeuroPoly for visual inspection of MRI images and manual segmentation of MS lesions. Therefore, naturally, the first step of the lesion segmentation process is to complete the FSLeyes tutorial ([FSLeyes documentation](https://open.win.ox.ac.uk/pages/fsl/fsleyes/fsleyes/userdoc/) and [video tutorial](https://www.youtube.com/playlist?list=PLIQIswOrUH69qFMNg8KYkEGkvCNEwlnfT)). Trainees are encouraged to learn keyboard shortcuts (ctrl+F to toggle an image, shift+↑ to scroll through volumes, ...).

Furthermore, it is recommended to get familiar with SCT for creating QCs and for manual correction ([SCT tutorial](https://spinalcordtoolbox.com/user_section/tutorials.html)). 

> ## More details:

Reviewer 2: ```suggestion
## Taxonomy to evaluation lesion segmentation
```

Author: Done !

```  

- If you are not sure about the segmentation on a subject, it should be flagged on GitHub for a more open discussion: here are some examples [(1)](https://github.com/ivadomed/ms-lesion-agnostic/issues/4#issuecomment-1947326493) and [(2)](https://github.com/ivadomed/ms-lesion-agnostic/issues/4#issuecomment-1947338624)

## Step 1: Get familiar with FSLeyes and SCT:
It is common practice to use FSLeyes at NeuroPoly for visual inspection of MRI images and manual segmentation of MS lesions. Therefore, naturally, the first step of the lesion segmentation process is to complete the FSLeyes tutorial ([FSLeyes documentation](https://open.win.ox.ac.uk/pages/fsl/fsleyes/fsleyes/userdoc/) and [video tutorial](https://www.youtube.com/playlist?list=PLIQIswOrUH69qFMNg8KYkEGkvCNEwlnfT)). Trainees are encouraged to learn keyboard shortcuts (ctrl+F to toggle an image, shift+↑ to scroll through volumes, ...).

Furthermore, it is recommended to get familiar with SCT for creating QCs and for manual correction ([SCT tutorial](https://spinalcordtoolbox.com/user_section/tutorials.html)). 

## Step 2: Spinal cord anatomy and lesion segmentation

> Before, moving on to MS lesion segmentation, trainees are advised to study the neuroanatomical structures of healthy spinal cords. Trainees should look at healthy spinal cords in MRI images of different contrasts: T2w, T1w, PSIR, STIR, MP2RAGE... A public dataset will be built for this purpose.

Author: SpineGeneric could be used for this purpose

To learn the specificity of MS lesions, trainees should work on differentiating MS lesions from other diseases. 

> One of the most challenging task of MS lesion segmentation is to distinguish the border of a lesion and the cerebrospinal fluid (CSF). To learn where to draw the lesion border, a set of tricky examples validated by a NeuroRadioligist will be created. 

Author: Which data should be used? MSSEG data? Ask a collaborator to help us on this?

It is common practice to use FSLeyes at NeuroPoly for visual inspection of MRI images and manual segmentation of MS lesions. Therefore, naturally, the first step of the lesion segmentation process is to complete the FSLeyes tutorial ([FSLeyes documentation](https://open.win.ox.ac.uk/pages/fsl/fsleyes/fsleyes/userdoc/) and [video tutorial](https://www.youtube.com/playlist?list=PLIQIswOrUH69qFMNg8KYkEGkvCNEwlnfT)). Trainees are encouraged to learn keyboard shortcuts (ctrl+F to toggle an image, shift+↑ to scroll through volumes, ...).

Furthermore, it is recommended to get familiar with SCT for creating QCs and for manual correction ([SCT tutorial](https://spinalcordtoolbox.com/user_section/tutorials.html)). 

## Step 2: Spinal cord anatomy and lesion segmentation
Before, moving on to MS lesion segmentation, trainees are advised to study the neuroanatomical structures of healthy spinal cords. Trainees should look at healthy spinal cords in MRI images of different contrasts: T2w, T1w, PSIR, STIR, MP2RAGE... A public dataset will be built for this purpose.

To learn the specificity of MS lesions, trainees should work on differentiating MS lesions from other diseases. 

> One of the most challenging tasks of MS lesion segmentation is to distinguish the border of a lesion and the cerebrospinal fluid (CSF). To learn where to draw the lesion border, a set of tricky examples validated by a NeuroRadioligist will be created. 

Author: Reference the videos made by Julien during work with Michelle Chen regarding MS lesion segmentation
